graphql/src/execution/cache.rs

```diff
@@ -75,3 +78,162 @@ impl<R: CheapClone> QueryCache<R> {
+            self.weight += weight;
+            self.cache.insert(key, (value, AtomicU64::new(0)));
+        }
+        fits_in_cache
+    }
+}
+
+/// Organize block caches by network names. Since different networks
+/// will be at different block heights, we need to keep their `CacheByBlock`
+/// separate
+pub struct QueryBlockCache {
+    shard: u8,
+    cache_by_network: Vec<(String, VecDeque<CacheByBlock>)>,
+    max_weight: usize,
+    max_blocks: usize,
+}
+
+impl QueryBlockCache {
+    pub fn new(max_blocks: usize, shard: u8, max_weight: usize) -> Self {
+        QueryBlockCache {
+            shard,
+            cache_by_network: Vec::new(),
+            max_weight,
+            max_blocks,
+        }
+    }
+
+    pub fn insert(
+        &mut self,
+        network: &str,
+        block_ptr: EthereumBlockPointer,
+        key: QueryHash,
+        result: Arc<QueryResult>,
+        weight: usize,
+        logger: Logger,
+    ) -> bool {
+        // Check if the cache is disabled
+        if self.max_blocks == 0 {
+            return false;
+        }
+
+        // Get or insert the cache for this network.
+        let cache = match self
+            .cache_by_network
+            .iter_mut()
+            .find(|(n, _)| n == network)
+            .map(|(_, c)| c)
+        {
+            Some(c) => c,
+            None => {
+                self.cache_by_network
+                    .push((network.to_owned(), VecDeque::new()));
+                &mut self.cache_by_network.last_mut().unwrap().1
+            }
+        };
+
+        // If there is already a cache by the block of this query, just add it there.
+        if let Some(cache_by_block) = cache.iter_mut().find(|c| c.block == block_ptr) {
+            return cache_by_block.insert(key, result.cheap_clone(), weight);
+        }
+
+        // We're creating a new `CacheByBlock` if:
+        // - There are none yet, this is the first query being cached, or
+        // - `block_ptr` is of higher or equal number than the most recent block in the cache.
+        // Otherwise this is a historical query that does not belong in the block cache.
+        if let Some(highest) = cache.iter().next() {
+            if highest.block.number > block_ptr.number {
+                return false;
+            }
+        };
+
+        if cache.len() == self.max_blocks {
+            // At capacity, so pop the oldest block.
+            // Stats are reported in a task since we don't need the lock for it.
+            let block = cache.pop_back().unwrap();
+            let shard = self.shard;
+            graph::spawn(async move {
+                let mut dead_inserts = 0;
+                let mut total_hits = 0;
+                for (_, hits) in block.cache.values() {
+                    let hits = hits.load(Ordering::SeqCst);
+                    total_hits += hits;
+                    if hits == 0 {
+                        dead_inserts += 1;
+                    }
+                }
```

Comment:
If we want to switch to caching on the second cache access, it would be good to know how many entries we have that are accessed only once, since those would go from hits to misses.

Reply:
Those would have a new status, `PreInsert` or something, we'll figure it out when we do that.